whelktool/scripts/cleanups/2019/04/remove-duplicated/from-list-move-items-and-delete-instance-duplicates.groovy

```diff
@@ -0,0 +1,70 @@
+/*
+ * This script moves holdings from one instance to another specified by
+ * the file id_lxl-2281_eboksberikning.txt The holdings are moved from
+ * the instance in the right column to the instance in the left column.
+ *
+ * See LXL-2380 for more info.
+ *
+ */
+
+PrintWriter failedBibIDs = getReportWriter("failed-to-delete-bibIDs")
+PrintWriter scheduledForChange = getReportWriter("scheduledForChange")
+PrintWriter instances = getReportWriter("instances")
+
+def file = new File(scriptDir, "id_lxl-2281_eboksberikning.txt");
+
+file.eachLine { line ->
+    def (goodInstanceId, badInstanceId) = line.split(";").collect { it.trim() }
+
+    def fullGoodInstanceId = "http://libris.kb.se/bib/${goodInstanceId}" as String
+
+    //Create a map of sigels mapping to items held by the good instance
+    def sigelToGoodHoldingMap = [:]
+
+    selectByIds([fullGoodInstanceId]) { goodBib ->
+        instances.println("Good bib ${fullGoodInstanceId} is a ${goodBib.graph[1][TYPE]}")
+        selectBySqlWhere("""
+                data#>>'{@graph,1,itemOf,@id}' = '${goodBib.graph[1][ID]}'
+        """, silent: true) { hold ->
+            sigelToGoodHoldingMap.put(hold.graph[1].heldBy[ID], hold)
+        }
+    }
+
+    selectByIds([badInstanceId]) { badInstance ->
+        instances.println("Bad bib ${badInstanceId} is a ${badInstance.graph[1][TYPE]}")
+        def fullBadInstanceId = badInstance.graph[1][ID]
+
+        selectBySqlWhere("""
+                data#>>'{@graph,1,itemOf,@id}' = '${fullBadInstanceId}'
+        """, silent: true) { hold ->
+
+            //If the sigel holding this item also has a holding on the good instance
+            //then delete the holding on the good instance before moving all holdings
+            def sigel = hold.graph[1].heldBy[ID]
+
+            if (sigelToGoodHoldingMap.containsKey(sigel)) {
+                def holdingToDelete = sigelToGoodHoldingMap.get(sigel)
+                def holdingToDeleteId = holdingToDelete.graph[1][ID]
+                holdingToDelete.scheduleDelete(onError: { e ->
+                    failedBibIDs.println("Failed to delete ${holdingToDeleteId} due to: $e")
+                }, loud: true)
+                scheduledForChange.println "For item held by ${sigel}:"
```

Comment:
If this script will be run threaded, having the prints on two separate lines can make the output jumbled. Better to combine them into one line instead.

Reply:
Added this!